Improve latest release automation

wadells opened the pull request:
## Summary
This changeset includes a couple changes:
1. I switch from git + shell to golang to calculate the most recent release for v6, v7, and v8 docker images, fixing https://github.com/gravitational/teleport/issues/9494
2. I fix a bug where we used lexicographic ordering instead of semver ordering when deciding to publish to apt repos (e.g. `v8.1.10` wouldn't be published because it sorted before `v8.1.9`)
3. Refactoring & tests to accommodate these changes

Fixes https://github.com/gravitational/teleport/issues/9494

The `version-ops/cmd/check` bugfixes need backporting to v8, so I'll probably take the entire changeset back.

## Testing Done
See the test cases added.  These all pass:

```
walt@work:~/git/teleport/build.assets/tooling$ go test ./...
ok      github.com/gravitational/teleport/build.assets/tooling/cmd/check        (cached)
ok      github.com/gravitational/teleport/build.assets/tooling/cmd/query-latest (cached)
?       github.com/gravitational/teleport/build.assets/tooling/cmd/render-tests [no test files]
?       github.com/gravitational/teleport/build.assets/tooling/lib/github       [no test files]
```

For integration with drone, I stripped out the `docker push`, and added build trigger for my branch. See the provisional `.drone.yml` change here:

https://github.com/gravitational/teleport/compare/walt/drone-docker-test#diff-b54b39f1afced2465e1f3641db9d5bbf4f3a7fcf890996dfedd3c197bcb7f8c7

With this in place, I got the following run:

https://drone.platform.teleport.sh/gravitational/teleport/10315/1/3 (updated 2022-02-16)

See the debug output at the end of the first step for the proof:
```
+ for FILE in /go/build/*.txt; do echo $FILE; cat $FILE; done
/go/build/CURRENT_VERSION_TAG.txt
v8.2.0
/go/build/CURRENT_VERSION_TAG_GENERIC.txt
8
/go/build/PREVIOUS_VERSION_ONE_TAG.txt
v7.3.13
/go/build/PREVIOUS_VERSION_ONE_TAG_GENERIC.txt
7
/go/build/PREVIOUS_VERSION_TWO_TAG.txt
v6.2.26
/go/build/PREVIOUS_VERSION_TWO_TAG_GENERIC.txt
6
```

vs a old, failing build:

https://drone.platform.teleport.sh/gravitational/teleport/10172/1/2

```
 for FILE in /go/build/*.txt; do echo $FILE; cat $FILE; done
/go/build/CURRENT_VERSION_TAG.txt
v8.2.0
/go/build/CURRENT_VERSION_TAG_GENERIC.txt
8
/go/build/PREVIOUS_VERSION_ONE_TAG.txt
v7.3.13^{}
/go/build/PREVIOUS_VERSION_ONE_TAG_GENERIC.txt
7
/go/build/PREVIOUS_VERSION_TWO_TAG.txt
v6.2.26
/go/build/PREVIOUS_VERSION_TWO_TAG_GENERIC.txt
6
```

wadells committed:
Integrate version-check into build.assets/tooling

This is a unified home as suggested by Trent here:

  https://github.com/gravitational/teleport/pull/10295#discussion_r807499882

Furthermore, I've split cmd code from lib code, in preparation for a new
command that will reuse the library code.

wadells committed:
Exclude draft releases from latest version logic

These should not be factored in when checking for the latest release
when we decide if we should release apt packages.

This also fixes a bug in sorting logic, where we were sorting
lexigraphically instead of by semver.

wadells committed:
Add an lexicographic test case

We are failing to sort properly when "9" is compared to "10".

wadells committed:
Fix release sorting

9 was comparing greater than 10, due to use of lexographic sorting

This would cause us to fail to publish apt packages when we roll over to
a patch release > 9.

wadells committed:
Add a command to query the latest release

This gives us a robust way to find the latest published release for a
Major or Major.Minor version.  This logic is useful for our automation
that publishes up-to-date teleport:X docker images

Contributes to https://github.com/gravitational/teleport/issues/9494

wadells committed:
Switch shell to golang for latest version detection (#10295)

The shell version detection is fragile and relatively error prone,
hopefully this go version will be more robust.  At the very least, it
will fix an issue where signed tags were creating incorrect download
urls.

Contributes to https://github.com/gravitational/teleport/issues/9494

webvictim reviewed (commented):
Drone parts LGTM on the whole, but I'll leave people with more Go chops to review the substantive changes. In general I always like the idea of moving away from my janky shell hacks.

I'd be careful with the rename from `version-check` to `version-ops` as this is probably hardcoded in a few places. It seems the linter is already angry :)

wadells commented:
> I'd be careful with the rename from `version-check` to `version-ops` as this is probably hardcoded in a few places. It seems the linter is already angry :)

I didn't realize we were even linting that until CI caught it. Thanks @timothyb89!  I `rg`'d for all other instances of `version-check` and make sure they were updated in the new 21a1485.

zmb3 commented on build.assets/version-ops/cmd/check/main.go:
```diff
@@ -24,15 +24,13 @@ import (
 
 	"golang.org/x/mod/semver"
 
+	"github.com/gravitational/teleport/build.assets/version-ops/lib"
```
`lib` is kind of an awkward name for a package. Doesn't tell me anything about what it is or where it comes from. Can we think of a more suitable name?

wadells replied:
Changed to `lib/github` in cb0ef079ea9c9e18873a662312f96b2a133126e0.

zmb3 commented on build.assets/version-ops/cmd/check/main_test.go:
```diff
@@ -121,6 +132,11 @@ type fakeGitHub struct {
+func (f *fakeGitHub) ListReleases(ctx context.Context, organization, repository string) ([]github.RepositoryRelease, error) {
+	ghReleases := make([]github.RepositoryRelease, 0)
+	for _, r := range f.releases {
+		tag := r
```
There's no need for the temporary `tag` variable here, is there?

`r` will be a copy of the string on each iteration, so `TagName: &r` seems fine to me.

zmb3 commented on build.assets/version-ops/cmd/query-latest/main.go:
```diff
@@ -0,0 +1,103 @@
+// query-latest ignores drafts and pre-releases.
+//
+// For example:
+//  query-latest -version-spec v8.1.5     -> v8.1.5
```
We don't really need a flag here. This seems easier to me, what do you think?

```
$ query-latest v8.1.5
```

wadells replied:
Yeah -- the flag was pattern matching with what @russjones initially implemented in https://github.com/gravitational/teleport/pull/9236/files#diff-05bfff32689895234ec9c80d4b2e6bc10792758e23d3ae1f29f2ad3d0300c60fR31-R39

I'm ok leaving it as a flag for now -- it is verbose but largely moot.

wadells replied:
I looped back and made this change in 5cac414.

zmb3 commented on build.assets/version-ops/cmd/query-latest/main.go:
```diff
@@ -0,0 +1,103 @@
+		return "", trace.Wrap(err)
+	}
+	if len(releases) == 0 {
+		return "", trace.BadParameter("failed to find any releases on GitHub")
```
Would `trace.NotFound` be more appropriate?

zmb3 commented on build.assets/version-ops/cmd/query-latest/main_test.go:
```diff
@@ -0,0 +1,110 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package main
```
Add an extra newline between license and package clause, otherwise `go doc` will render the license as documentation.

wadells replied:
Done in affbe4648a7cc9511ff11ce690c161333d71461d, b40925a19dfaa2f00050d3258b9fbc7333b9f686, c32db91716fc6e9bbe06a4efad0aeaa460105bda.

zmb3 commented on build.assets/version-ops/cmd/query-latest/main_test.go:
```diff
@@ -0,0 +1,110 @@
+				"v8.1.10",
+				"v8.0.11",
+			},
+			wantErr: false,
```
If you want to use `testify/require`, you could clean this up a bit. Instead of having `wantErr bool`, you could do something like `assert require.ErrorAssertionFunc`, and each test case would either pass `require.Error` or `require.NoError`.

Then in your loop, no need for an if-statement, just `test.assert(t, err)`

If you want to minimize dependencies, what you have here is totally fine, though you're already pulling in testify so it's not even a net new dependency.

wadells replied:
Thanks -- I'll check out testify in the future.  I haven't worked with it much yet and it is orthogonal to the purpose of this patch, so I'm gonna decline this for now.

tcsc replied:
Ah, I said the same thing above. Fair enough response.

wadells replied:
Two folks saying the same thing adds some weight. :smile:  Fixed in 6e10e88.

zmb3 reviewed (approved):
Added a few suggestions, but feel free to ignore any or all of them. This looks good.

wadells commented on build.assets/version-ops/cmd/check/main.go:
```diff
@@ -78,20 +76,27 @@ func parseFlags() (string, string, error) {
 
-	if semver.Compare(tag, releases[0]) <= 0 {
-		return trace.BadParameter("found newer version of release, not releasing. Latest release: %v, tag: %v", releases[0], tag)
+	tags := make([]string, 0)
```
Changed in 708bb8d9a799d3df72f2ea02dd8fc26f7901824c, 5c7e8d97ae2389baa45d8057f8b10eb330988a18.  I'll squash these later.

tcsc commented on build.assets/version-ops/cmd/check/main.go:
```diff
@@ -24,15 +24,13 @@ import (
 
 	"golang.org/x/mod/semver"
 
+	"github.com/gravitational/teleport/build.assets/version-ops/lib/github"
```
I've started a  module in `build.assets/tooling` for build-related tooling (basically anything used from the Makefile down). That may (or may not be) the right place for this.

wadells replied:
Moved to `build.assets/tooling` in 1605a41.

tcsc commented on build.assets/version-ops/cmd/check/main_test.go:
```diff
@@ -13,11 +13,14 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 	"context"
 	"testing"
+
+	"github.com/google/go-github/v41/github"
```
If you're up for including it, the rest of teleport is using Testify for tests (`github.com/stretchr/testify`), which may simplify your test  cases a bit.

```go
tests := []struct {
		desc     string
		tag      string
		releases []string
		wantErr  require.ErrorAssertionFunc
	}{
		{
			desc:    "fail-rc",
			tag:     "v9.0.0-rc.1",
			wantErr: require.Error
		},
		//...
		{
			desc:    "pass",
			tag:     "v8.0.1",
			wantErr: require.NoError,
		},
	}
	for _, test := range tests {
		t.Run(test.desc, func(t *testing.T) {
			test.wantError(t, checkPrerelease(test.tag))
		})
	}
}
```

Not a huge win for one test, but it makes everything have the same vocabulary

wadells replied:
[2nd mention](https://github.com/gravitational/teleport/pull/10295#discussion_r805071270) is a charm. :smile:  Done in 6e10e88.

tcsc commented on build.assets/version-ops/cmd/query-latest/main.go:
```diff
@@ -0,0 +1,103 @@
+
+	// filter drafts and prereleases, which shouldn't be tracked by latest docker images
+	var tags []string
+	for _, r := range releases {
```
I've seen this pattern in a few places. Is it worth pulling out into a function? E.g.

```go
func getFilteredTagsreleases []*Release, predicate func(*Release) bool) []tag { /*...*/ }

// ..

tags := getFilteredTags(releases, func(r *Release) bool { 
    return !(r.GetDraft() && r.GetPrerelease())
})
```
too much? The individual sites might be different enough to make this impractical.

wadells replied:
I'd prefer to leave it duplicated for the moment, especially since the two instances have slightly different behavior.

The difference in behavior is:

A more recent pre-release will block publishing apt packages (e.g. do not publish 8.2.0 apt packages if 8.2.1-dev.1 is a github release). This is because we don't want to downgrade apt versions *ever*.  Thus we don't filter pre-releases from consideration -- on the off chance a pre-release escaped.

For publishing latest versions, we do filter pre-releases from consideration, because we don't want our "latest" docker images to pick up pre-releases.  Thus we do filter pre-releases from consideration when determining the latest release on a branch for docker image purposes.

Honestly, this is a pretty slim distinction, especially since we block pre-releases from being published to the apt repo.

wadells committed:
Switch to testify

This saves us a couple lines of code and is a consistent review
recommendation. Better to learn it myself than keep pushing back. :)

wadells commented:
I'm gonna re-run integration testing, since there have been several path-fiddly changes that could trip up the drone.  I'll reenable auto-merge and try to get this in once I've completed the new round of testing.

tcsc commented on build.assets/tooling/cmd/query-latest/main_test.go:
```diff
@@ -0,0 +1,107 @@
+			}
+			latest, err := getLatest(context.Background(), test.spec, gh)
+			test.wantErr(t, err)
+			if test.latest != latest {
```
```
require.Equal(t, test.latest, latest)
```

wadells replied:
Thanks. Fixed in [5682e98](https://github.com/gravitational/teleport/pull/10295/commits/5682e98b4d01f0fc3bf14b6b7450b322ab2401b0).

wadells commented:
> I'm gonna re-run integration testing, since there have been several path-fiddly changes that could trip up the drone. I'll reenable auto-merge and try to get this in once I've completed the new round of testing.

Integration ran cleanly: https://drone.platform.teleport.sh/gravitational/teleport/10315/1/3. Diving into the merge free-for-all :)
